stream_alert/classifier/payload/s3.py

```diff
@@ -0,0 +1,214 @@
+                                 'from S3: {}'.format(self.bucket, self.key, self.display_size))
+
+    @staticmethod
+    def _shred_temp_directory():
+        """Delete all objects in the container's temp directory"""
+        LOGGER.debug('Shredding temp directory')
+
+        for root, dirs, files in os.walk(tempfile.gettempdir(), topdown=False):
+            for name in files:
+                subprocess.check_call([  # nosec
+                    'shred', '--force', '--iterations=1',
+                    '--remove', os.path.join(root, name)])
+            for name in dirs:
+                os.rmdir(os.path.join(root, name))  # nosec
+
+    @staticmethod
+    def _gz_reader(open_file):
+        open_file.seek(0)
+        reader = gzip.GzipFile(fileobj=open_file, mode='r')
+        try:
+            # Test to ensure this is gzip data, then rewind
+            reader.read(1)
+            reader.rewind()
+        except IOError:
+            # Fall back on the default reader
+            reader = open_file
+            reader.seek(0)
+
+        # Return either the gzip reader or the original reader
+        return reader
+
+    @staticmethod
+    def _jsonlines_reader(open_file):
+        open_file.seek(0)
+        reader = None
+        try:
+            json_lines = jsonlines.Reader(open_file)
+            json_lines.read()
+            reader = json_lines
+        except ValueError:
+            reader = open_file
+
+        open_file.seek(0)
+        return reader
+
+    @classmethod
+    def _read_downloaded_object(cls, open_file):
+        """Read the contents of the downloaded S3 file using the open file handle
+
+        Args:
+            open_file (file-like object): File handle from which to read the data
+
+        Yields:
+            tuple: line number, contents of the line being read
+        """
+        reader = cls._gz_reader(open_file)
+        try:
+            # Try to just load the reader as regular json first
+            yield 1, json.load(reader)
+            return
+        except ValueError:
+            pass
+
+        # Iterate over the lines, returning each
+        # This could be dicts from a jsonlines.Reader, or raw strings
+        for line_num, line in enumerate(cls._jsonlines_reader(reader), start=1):
+            yield line_num, line.strip() if isinstance(line, basestring) else line
+
+    def _read_file(self):
+        """Download and read the contents of the S3 file
+
+        Yields:
+            tuple: line number, contents of the line being read
+        """
+        # Shred the temp dir before downloading
+        self._shred_temp_directory()
+        bucket = self._unquote(self.bucket)
+        key = self._unquote(self.key)
+
+        # Use tempfile.TemporaryFile to do the download
+        # This will automatically close/get garbage collected upon completion
+        with tempfile.TemporaryFile() as download:
```

Comment:
right now it looks like you're only shredding before actually writing the sensitive data.
i might use a NamedTemporaryFile and shred it immediately after consumption (over shredding the directory at the end which may not run if the lambda terminates early)

Reply:
good point regarding shredding after, not before. however, I'm confused what the benefit of using `NamedTemporaryFile` would be?